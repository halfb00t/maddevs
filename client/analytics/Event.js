/* eslint-disable no-console, class-methods-use-this, no-underscore-dangle */
import uniqid from '~/helpers/generatorUid'

const LOCAL_STORAGE_KEYS = {
  ID: 'GA_USER_ID',
  TYPE: 'GA_USER_TYPE',
}

export const addUserType = type => {
  // TYPE: 'hr_candidate' | 'lead' | 'download_ebook' | undefined
  localStorage.setItem(LOCAL_STORAGE_KEYS.TYPE, type)
}

export class AnalyticsEvent {
  constructor(action, strict = true) {
    if (!action) {
      throw new Error(`Event action is missing for ${this.action}, please add an action for object`)
    }
    this.strict = strict
    this.action = action
    this.properties = {}
    this._applyUser()
  }

  _applyUser() {
    if ('window' in global) {
      const existId = localStorage.getItem(LOCAL_STORAGE_KEYS.ID)
      const type = localStorage.getItem(LOCAL_STORAGE_KEYS.TYPE)

      const userId = existId || uniqid()

      localStorage.setItem(LOCAL_STORAGE_KEYS.ID, userId)

      this.properties.user_id = userId
      this.properties.user_type = type
    }
  }

  _handleError(message) {
    if (this.strict) throw new Error(message)
    console.error(message)
  }

  setCategory(category) {
    if (!category) {
      this._handleError('Event category is required')
    }
    this.properties.event_category = category
    return this
  }

  setField(name, value) {
    if (!name || !(value !== undefined && value !== null)) {
      this._handleError('AnalyticsEvent.setField() method is failed. Please add name or values params')
    }
    this.properties[name] = value
    return this
  }

  _setPath() {
    this.properties.path = window.location.pathname
  }

  _collectGoogleAnalyticsKeys() {
    const keys = window.dataLayer
      ?.filter(dataLayerValue => dataLayerValue[0] === 'config' && dataLayerValue.length >= 3)
      .map(dataLayerValue => dataLayerValue[1]) || []

    return Array.from(new Set(keys))
  }

  _log(analyticsKeys) {
    const msg = [
      '<------- ANALYTICS EVENT SENDING -------->',
      `EVENT: ${this.action}`,
      `PROPS: ${JSON.stringify(this.properties)}`,
      `ANALYTICS: ${JSON.stringify(analyticsKeys)}`,
    ]

    console.log(msg.join('\n'))
  }

  send() {
    if (process.env.domain === 'https://maddevs.io' || process.env.domain === 'https://maddevs.co') {
      this._setPath()
      this._applyUser()

      const analyticsKeys = this._collectGoogleAnalyticsKeys()

      if (process.env.NODE_ENV === 'development') {
        this._log(analyticsKeys)
      }

      analyticsKeys.forEach(analyticsId => {
        const properties = { ...this.properties, send_to: analyticsId }
        try {
          window.gtag('event', this.action, properties)
        } catch (error) {
          this._handleError(error)
        }
      })
    }
  }
}
