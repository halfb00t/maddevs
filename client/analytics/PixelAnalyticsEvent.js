/* eslint-disable no-console, class-methods-use-this, no-underscore-dangle */
import uniqid from '~/helpers/generatorUid'

const LOCAL_STORAGE_KEYS = {
  ID: 'GA_USER_ID',
  TYPE: 'GA_USER_TYPE',
}

export const TRACK_TYPES = {
  TRACK: 'track',
  CUSTOM_TRACK: 'trackCustom',
}

export const addUserType = type => {
  // TYPE: 'hr_candidate' | 'lead' | 'download_ebook' | undefined
  localStorage.setItem(LOCAL_STORAGE_KEYS.TYPE, type)
}

export class PixelAnalyticsEvent {
  constructor(action, trackType = TRACK_TYPES.TRACK, strict = true) {
    if (!action) {
      throw new Error(`Event action is missing for ${this.action}, please add an action for object`)
    }
    this.strict = strict
    this.action = action
    this.properties = {}
    this.trackType = trackType
    this._applyUser()
  }

  _applyUser() {
    if ('window' in global) {
      const existId = localStorage.getItem(LOCAL_STORAGE_KEYS.ID)
      const type = localStorage.getItem(LOCAL_STORAGE_KEYS.TYPE)

      const userId = existId || uniqid()

      localStorage.setItem(LOCAL_STORAGE_KEYS.ID, userId)

      this.properties.user_id = userId
      this.properties.user_type = type
    }
  }

  _handleError(message) {
    if (this.strict) throw new Error(message)
    console.error(message)
  }

  setCategory(category) {
    if (!category) {
      this._handleError('Event category is required')
    }
    this.properties.event_category = category
    return this
  }

  setField(name, value) {
    if (!name || !(value !== undefined && value !== null)) {
      this._handleError('AnalyticsEvent.setField() method is failed. Please add name or values params')
    }
    this.properties[name] = value
    return this
  }

  _setPath() {
    this.properties.path = window.location.pathname
  }

  _collectGoogleAnalyticsKeys() {
    const keys = window.dataLayer
      ?.filter(dataLayerValue => dataLayerValue[0] === 'config' && dataLayerValue.length >= 3)
      .map(dataLayerValue => dataLayerValue[1]) || []

    return Array.from(new Set(keys))
  }

  _log(analyticsKeys) {
    const msg = [
      '<------- ANALYTICS EVENT SENDING -------->',
      `EVENT: ${this.action}`,
      `PROPS: ${JSON.stringify(this.properties)}`,
      `ANALYTICS: ${JSON.stringify(analyticsKeys)}`,
    ]

    console.log(msg.join('\n'))
  }

  send() {
    if (window.location.hostname === 'maddevs.io'
      || window.location.hostname === 'maddevs.co'
      || window.location.origin === 'https://maddevs.co'
      || window.location.origin === 'https://maddevs.io'
    ) {
      this._setPath()
      this._applyUser()

      const analyticsKeys = this._collectGoogleAnalyticsKeys()

      if (process.env.NODE_ENV === 'development') {
        this._log(analyticsKeys)
      }

      analyticsKeys.forEach(analyticsId => {
        const properties = { ...this.properties, send_to: analyticsId }
        try {
          window.fbq(this.trackType, this.action, properties)
        } catch (error) {
          this._handleError(error)
        }
      })
    }
  }
}
