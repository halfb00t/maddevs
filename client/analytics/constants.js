export const CONTENT_GROUPS = [
  { url: ['http://maddevs.io/'], name: 'home_page' },
  { url: ['https://maddevs.io/case-studies/'], name: 'case_studies' },
  { url: ['https://maddevs.io/case-studies/clutch/'], name: 'case' },
  { url: ['https://maddevs.io/blog/'], name: 'blog' },
  { url: ['https://maddevs.io/ebooks/pricing-strategies/'], name: 'pricing-strategies' },
  { url: ['https://maddevs.io/customer-university/custom-software-development-pricing-strategies/'], name: 'cu_post' },
  { url: ['https://maddevs.io/blog/automation-of-customer-service/'], name: 'blog_post' },
  { url: ['https://maddevs.io/open-source/'], name: 'open_source' },
  {
    url: [
      'https://maddevs.io/services/',
      'https://maddevs.io/software-testing/',
    ],
    name: 'services',
  },
  {
    url: [
      'https://maddevs.io/development-process/',
      'https://maddevs.io/delivery-models/',
    ],
    name: 'approach',
  },
  { url: ['https://maddevs.io/team/'], name: 'team' },
  { url: ['https://maddevs.io/careers/'], name: 'careers' },
  { url: ['https://maddevs.io/ru/careers/senior-golang/'], name: 'job_post' },
  { url: ['https://maddevs.io/contact-us/'], name: 'contact_us' },
  { url: ['https://maddevs.io/tag/frontend-development/'], name: 'tag' },
  { url: ['https://maddevs.io/author/denis-grushkin/'], name: 'author' },
  {
    url: [
      'https://maddevs.io/privacy/',
      'https://maddevs.io/sustainability-policy/',
      'https://maddevs.io/gdpr/',
      'https://maddevs.io/nda/',
      'https://maddevs.io/faq/',
    ],
    name: 'policy',
  },
  {
    url: [
      'https://maddevs.io/transportation/',
      'https://maddevs.io/blockchain/',
    ],
    name: 'industries',
  },
]

export const EVENT_CATEGORIES = {
  CLICK: 'click',
  SCROLL: 'scroll',
  SUBMIT: 'submit',
}

export const ANALYTICS_IDS = {
  GA4: process.env.analytics4Key,
  UNIVERSAL: process.env.analyticsUniversalKey,
}
