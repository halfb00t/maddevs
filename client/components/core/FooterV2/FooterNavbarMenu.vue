<template>
  <div
    v-if="menuRoutes && menuRoutes.length"
    class="footer-menu"
  >
    <h5
      v-if="menuName"
      class="footer-menu__name"
    >
      {{ menuName }}
    </h5>
    <ul class="footer-menu__list">
      <li
        v-for="{ label, link } in menuRoutes"
        :key="label"
        class="footer-menu__item"
      >
        <div
          v-if="label"
        >
          <NuxtLink
            v-if="!extractMenuLink(link).isExternalLink"
            :to="extractMenuLink(link).url"
            data-testid="menu-item-internal-link"
            class="footer-menu__item-link"
            @click.native="goToPage"
          >
            {{ label }}
          </NuxtLink>
          <a
            v-else
            :href="extractMenuLink(link).url"
            data-testid="menu-item-external-link"
            target="_blank"
          >
            {{ label }}
          </a>
        </div>
      </li>
    </ul>
  </div>
</template>

<script>
import extractMenuLink from '@/helpers/extractMenuLink'

export default {
  name: 'FooterNavbarMenu',

  props: {
    menuName: {
      type: String,
      default: '',
    },

    menuRoutes: {
      type: Array,
      default: () => ([]),
    },
  },

  methods: {
    extractMenuLink,

    goToPage() {
      this.$root.$emit('changed-page')
    },
  },
}
</script>

<style lang="scss" scoped>
.footer-menu__item a {
  color: $text-color--white-primary;
  font-size: 12px;
  text-transform: capitalize;
}

.footer-menu {
  margin-bottom: 30px;

  &__name {
    color: $text-color--grey-pale;
    font-size: 12px;
    margin-bottom: 15px;
    text-transform: capitalize;
    font-weight: normal;
  }

  &__item {
    margin-bottom: 8px;

    &-link {
      transition: color .2s;

      &:hover {
        color: $text-color--red;
      }
    }
  }
}
</style>
