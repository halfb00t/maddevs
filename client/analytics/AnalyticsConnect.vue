<template>
  <div />
</template>

<script>
import { ANALYTICS_IDS } from '@/analytics/constants'

export default {
  data() {
    return {
      ANALYTICS_IDS,
    }
  },

  mounted() {
    window.dataLayer = window.dataLayer || []

    function gtag(...args) {
      window.dataLayer.push(args)
    }

    window.gtag = gtag

    window.gtag('js', new Date())
    window.gtag('config', this.ANALYTICS_IDS.GA4, { debug_mode: true })
    window.gtag('config', this.ANALYTICS_IDS.UNIVERSAL, { debug_mode: true })
  },
}
</script>
