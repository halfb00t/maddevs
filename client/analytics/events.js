import { AnalyticsEvent } from './Event'
import { EVENT_CATEGORIES } from './constants'

export const headerClickEvent = new AnalyticsEvent('click').setCategory(EVENT_CATEGORIES.CLICK)

export const phoneClickEvent = new AnalyticsEvent('phone_click').setCategory(EVENT_CATEGORIES.CLICK)

export const emailClickEvent = new AnalyticsEvent('email_click').setCategory(EVENT_CATEGORIES.CLICK)

export const contactMeClickEvent = new AnalyticsEvent('contact_me_click').setCategory(EVENT_CATEGORIES.CLICK)

export const contactMeSubmitEvent = new AnalyticsEvent('submit_contact_me').setCategory(EVENT_CATEGORIES.SUBMIT)

export const cardClickEvent = new AnalyticsEvent('card_click').setCategory(EVENT_CATEGORIES.CLICK)

export const socialNetworkClickEvent = new AnalyticsEvent('social_network_click').setCategory(EVENT_CATEGORIES.CLICK)

export const footerFormSubmitEvent = new AnalyticsEvent('submit_footer_form').setCategory(EVENT_CATEGORIES.SUBMIT)

export const contactsClickEvent = new AnalyticsEvent('contacts_click').setCategory(EVENT_CATEGORIES.CLICK)

export const blockClickEvent = new AnalyticsEvent('blog_click').setCategory(EVENT_CATEGORIES.CLICK)

export const categoryPageClickEvent = new AnalyticsEvent('category_click').setCategory(EVENT_CATEGORIES.CLICK)

export const pressCenterClickEvent = new AnalyticsEvent('press_click').setCategory(EVENT_CATEGORIES.CLICK)

export const linkedinUserClickEvent = new AnalyticsEvent('linkedin_user_click').setCategory(EVENT_CATEGORIES.CLICK)

export const careersClickEvent = new AnalyticsEvent('careers_click').setCategory(EVENT_CATEGORIES.CLICK)

export const hrSiteClickEvent = new AnalyticsEvent('hr_site_click').setCategory(EVENT_CATEGORIES.CLICK)

export const githubClickEvent = new AnalyticsEvent('github_click').setCategory(EVENT_CATEGORIES.CLICK)

export const writeUsClickEvent = new AnalyticsEvent('write_us_click').setCategory(EVENT_CATEGORIES.CLICK)

export const submitCvFormEvent = new AnalyticsEvent('submit_cv').setCategory(EVENT_CATEGORIES.SUBMIT)

export const pricingStrategiesEbookSubmitFormEvent = new AnalyticsEvent('submit_pricing_strategies_ebook_form').setCategory(EVENT_CATEGORIES.SUBMIT)

export const approachDPEbookSubmitFormEvent = new AnalyticsEvent('submit_approach_dp_ebook_form').setCategory(EVENT_CATEGORIES.SUBMIT)
