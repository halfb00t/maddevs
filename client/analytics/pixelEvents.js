import { EVENT_CATEGORIES } from './constants'
import { PixelAnalyticsEvent, TRACK_TYPES } from './PixelAnalyticsEvent'

export const headerClickEvent = new PixelAnalyticsEvent('click').setCategory(EVENT_CATEGORIES.CLICK)

export const phoneClickPixelEvent = new PixelAnalyticsEvent('phone_click').setCategory(EVENT_CATEGORIES.CLICK)

export const contactMeClickPixelEvent = new PixelAnalyticsEvent('contact_me').setCategory(EVENT_CATEGORIES.CLICK)

export const emailClickPixelEvent = new PixelAnalyticsEvent('email_click').setCategory(EVENT_CATEGORIES.CLICK)

export const hrSiteClickPixelEvent = new PixelAnalyticsEvent('hr_site_click').setCategory(EVENT_CATEGORIES.CLICK)

export const socialNetworkClickPixelEvent = new PixelAnalyticsEvent('social_network_click').setCategory(EVENT_CATEGORIES.CLICK)

export const writeUsClickPixelEvent = new PixelAnalyticsEvent('write_us_click').setCategory(EVENT_CATEGORIES.CLICK)

export const ebookSubmitFormPixelEvent = new PixelAnalyticsEvent('submit_ebook_form').setCategory(EVENT_CATEGORIES.SUBMIT)

export const githubClickPixelEvent = new PixelAnalyticsEvent('github_click').setCategory(EVENT_CATEGORIES.CLICK)

export const pressCenterClickPixelEvent = new PixelAnalyticsEvent('press_click').setCategory(EVENT_CATEGORIES.CLICK)

export const submitCvFormPixelEvent = new PixelAnalyticsEvent('submit_cv').setCategory(EVENT_CATEGORIES.CLICK)

export const linkedinUserClickPixelEvent = new PixelAnalyticsEvent('linkedin_user_click').setCategory(EVENT_CATEGORIES.CLICK)

export const twitterNetworkPixelEvent = new PixelAnalyticsEvent('twitter', TRACK_TYPES.CUSTOM_TRACK).setCategory(EVENT_CATEGORIES.CLICK)

export const facebookNetworkPixelEvent = new PixelAnalyticsEvent('facebook', TRACK_TYPES.CUSTOM_TRACK).setCategory(EVENT_CATEGORIES.CLICK)

export const linkedinNetworkPixelEvent = new PixelAnalyticsEvent('facebook', TRACK_TYPES.CUSTOM_TRACK).setCategory(EVENT_CATEGORIES.CLICK)
