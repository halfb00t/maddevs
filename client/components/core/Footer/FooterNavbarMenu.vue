<template>
  <div
    v-if="menuRoutes && menuRoutes.length"
    class="footer-menu"
  >
    <h3
      v-if="menuName"
      class="footer-menu__name"
    >
      {{ menuName }}
    </h3>
    <ul class="footer-menu__list">
      <li
        v-for="{ label, link } in menuRoutes"
        :key="label"
        class="footer-menu__item"
      >
        <div
          v-if="label"
        >
          <NuxtLink
            v-if="!extractMenuLink(link).isExternalLink"
            :to="extractMenuLink(link).url"
            data-testid="menu-item-internal-link"
            class="footer-menu__item-link"
            @click.native="goToPage($event)"
          >
            {{ label }}
          </NuxtLink>
          <a
            v-else
            :href="extractMenuLink(link).url"
            data-testid="menu-item-external-link"
            target="_blank"
          >
            {{ label }}
          </a>
        </div>
      </li>
    </ul>
  </div>
</template>

<script>
import extractMenuLink from '@/helpers/extractMenuLink'
import { blockClickEvent, contactsClickEvent } from '@/analytics/events'

export default {
  name: 'FooterNavbarMenu',

  props: {
    menuName: {
      type: String,
      default: '',
    },

    menuRoutes: {
      type: Array,
      default: () => ([]),
    },
  },

  methods: {
    extractMenuLink,

    goToPage(event) {
      if (event.target.pathname === '/contact-us/') contactsClickEvent.send()
      if (event.target.pathname === '/blog/') blockClickEvent.send()
      this.$root.$emit('changed-page')
    },
  },
}
</script>

<style lang="scss" scoped>
.footer-menu__item a {
  color: $text-color--white-primary;
  font-size: 12px;
  text-transform: capitalize;
}

a.nuxt-link-exact-active {
  color: $text-color--red !important;

  &::before {
    opacity: 1;
  }
}

.footer-menu {
  margin-bottom: 30px;

  &__name {
    color: $text-color--grey-pale;
    font-size: 12px;
    margin-bottom: 15px;
    text-transform: capitalize;
    font-weight: normal;
  }

  &__item {
    margin-bottom: 8px;

    &-link {
      transition: color .2s;

      &:hover {
        color: $text-color--red;
      }
    }
  }
}
</style>
