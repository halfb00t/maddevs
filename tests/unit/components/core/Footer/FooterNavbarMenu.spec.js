import { render, fireEvent, screen } from '@testing-library/vue'
import { shallowMount } from '@vue/test-utils'
import { contactsClickEvent } from '@/analytics/events'
import FooterNavbarMenu from '@/components/core/Footer/FooterNavbarMenu'
import { footerNavigation } from './footerNavigationFixture'

jest.mock('~/helpers/generatorUid')

const contactsClick = jest.spyOn(contactsClickEvent, 'send')
  .mockImplementation(() => {
  })

const stubs = ['NuxtLink']

const props = {
  menuName: footerNavigation.company.menus[0].name,
  menuRoutes: footerNavigation.company.menus[0].routes,
}
describe('FooterNavbarColumn', () => {
  it('should correct render footer navbar menu', () => {
    const { container } = render(FooterNavbarMenu, {
      stubs,
      props,
    })
    expect(container)
      .toMatchSnapshot()
  })
  it('should correct render menuRoutes without data', () => {
    const wrapper = shallowMount(FooterNavbarMenu, {
      stubs,
      propsData: props,
    })
    expect(wrapper.vm.$options.props.menuRoutes.default.call())
      .toEqual([])
  })
  it('should correct render menuName without data', () => {
    const wrapper = shallowMount(FooterNavbarMenu, {
      stubs,
      propsData: props,
    })
    expect(wrapper.vm.$options.props.menuName.default)
      .toEqual('')
  })
  it('should correct send event from contact-us to analytic', async () => {
    render(FooterNavbarMenu, {
      propsData: props,
      stubs: {
        NuxtLink: { template: '<a href="https://maddevs.io/contact-us/"></a>' },
      },
    })
    const link = screen.getAllByTestId('menu-item-internal-link')
    await fireEvent.click(link[3])
    expect(contactsClick)
      .toHaveBeenCalledTimes(1)
  })
  it('should correct send event from blog to analytic', async () => {
    render(FooterNavbarMenu, {
      propsData: props,
      stubs: {
        NuxtLink: { template: '<a href="https://maddevs.io/blog/"></a>' },
      },
    })
    const link = screen.getAllByTestId('menu-item-internal-link')
    await fireEvent.click(link[3])
    expect(contactsClick)
      .toHaveBeenCalledTimes(1)
  })
})
