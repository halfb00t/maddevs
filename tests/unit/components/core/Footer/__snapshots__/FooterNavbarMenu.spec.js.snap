// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterNavbarColumn should correct render footer navbar menu 1`] = `
<div>
  <div
    class="footer-menu"
  >
    <h5
      class="footer-menu__name"
    >
      
    Company
  
    </h5>
     
    <ul
      class="footer-menu__list"
    >
      <li
        class="footer-menu__item"
      >
        <div>
          <nuxtlink-stub
            class="footer-menu__item-link"
            data-testid="menu-item-internal-link"
            to="/about/"
          >
            
          About Mad Devs
        
          </nuxtlink-stub>
        </div>
      </li>
      <li
        class="footer-menu__item"
      >
        <div>
          <nuxtlink-stub
            class="footer-menu__item-link"
            data-testid="menu-item-internal-link"
            to="/team/"
          >
            
          Team
        
          </nuxtlink-stub>
        </div>
      </li>
      <li
        class="footer-menu__item"
      >
        <div>
          <nuxtlink-stub
            class="footer-menu__item-link"
            data-testid="menu-item-internal-link"
            to="/careers/"
          >
            
          Careers
        
          </nuxtlink-stub>
        </div>
      </li>
      <li
        class="footer-menu__item"
      >
        <div>
          <nuxtlink-stub
            class="footer-menu__item-link"
            data-testid="menu-item-internal-link"
            to="/contact-us/"
          >
            
          Contacts
        
          </nuxtlink-stub>
        </div>
      </li>
      <li
        class="footer-menu__item"
      >
        <div>
          <nuxtlink-stub
            class="footer-menu__item-link"
            data-testid="menu-item-internal-link"
            to="/faq/"
          >
            
          FAQ
        
          </nuxtlink-stub>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
